rfcs/0001-extended-lookup-tables.md

```diff
@@ -0,0 +1,503 @@
+* allow random-access lookups in arrays as a base-primitive for user programs
+* allow circuits to use low-cost branching and other familiar programming
+  primitives
+  - this depends on the development of a bytecode interpreter on top of the
+    lookup-table primitives.
+* allow circuits to load and execute some 'user provided' bytecode
+  - e.g. to run a small bytecode program provided in a transaction at
+    block-production time.
+* allow zkOracles to load a full HTTPS transcript without using an expensive
+  in-circuit commitment.
+
+## Detailed design
+[detailed-design]: #detailed-design
+
+In order to support the desired goals, we first define 3 types of table:
+* **fixed tables** are tables declared as part of the constraint system, and
+  are the same for every proof of that circuit.
+* **runtime tables** are tables whose contents are determined at proving time,
+  to be used for array accesses to data from the witness.
+* **side-loaded tables** are tables that are committed to in a
+  proof-independent way, so that they may be reused across proofs and/or signed
+  without knowledge of the specific proof, but may be different for each
+  execution of the circuit.
+
+These distinct types of tables are a slight fiction: often it will be desirable
+to have some fixed part of runtime or side-loaded tables, e.g. to ensure that
+indexing is reliable. For example, a table representing an array of values
+`[x, y, z]` in the proof might be laid out as
+
+|value1| value2 |
+|------|--------|
+|0     |?runtime|
+|1     |?runtime|
+|2     |?runtime|
+
+where the `value1` entries are fixed in the constraint system. This ensure that
+a malicious prover is not free to create multiple values with the same index,
+such as `[(0,a), (0,x), (1,y), (2,z)]`, otherwise the value `a` might be used
+where the value `x` was intended.
+
+### Combining multiple tables
+
+The current implementation only supports a single table. In order to make
+multiple tables available, we either need to run a separate PLOOKUP argument
+for each table, or to concatenate the tables to form an combined table and
+identify the values from each table in such a way that the prover cannot use
+values from the 'wrong' table.
+
+We already solve a similar problem with 'vector' tables -- tables where a row
+contains multiple values -- where we want to ensure that the prover used an
+entry `(x, y, z)` from the table, and not some parts of different entries, or
+some other value entirely. In order to do this, we generate a randomising field
+element `joint_combiner`, where the randomness depends on the circuit witness,
+and then compute `x + y * joint_combiner + z * joint_combiner^2`. Notice that,
+if the prover wants to select some `(a, b, c)` not in the table and claim that
+it in fact was, they are not able to use the knowledge of the `joint_combiner`
+to choose their `(a, b, c)`, since their choice will change the value of the
+`joint_combiner`.
```

Comment:
I really like that these RFCs are opportunities to define terms that we can use to communicate more clearly (and even reuse in the implementation!), you're doing a lot of this here which I appreciate :) I suggest adding terms to distinguish the different types of look up tables between transformation. At this point I think these two terms would be really useful (in the implementation as well):

* lookup table (perhaps shortened as LT?) -> the actual table with columns
* combined lookup table (CLT) -> the table with a single column (after using `joint_combiner`)

Reply:
There are a few different styles of 'combining' at play here:
* we 'combine' the fixed tables into a single fixed table
* we 'combine' the fixed table with the runtime and/or side-loaded tables, where the latter are used as masks
* we 'combine' any of the 4 different tables above with the `table_id`s column
* we 'combine' the values in the table above with the `joint_combiner` to create a single value representing each row.

I'll try to think of and define some reasonable terms. At the moment, I'm leaning towards
* 'fixed tables'
* 'combined fixed table'
* 'fixed+runtime table'
* 'lookup table' (fixed+runtime+side-loaded)
* an 'identified' table is a table considered along with its `table_id`
  - e.g. 'identified lookup table', 'identified fixed+runtime table', etc.
* a 'collapsed' table is an 'identified' table where the rows are collapsed into a single value using the `joint_combiner`